app/services/billing-accounts/change-address.service.js

```diff
@@ -0,0 +1,340 @@
+  const addressInstance = _transformAddress(timestamp, address)
+  const companyInstance = _transformCompany(timestamp, agentCompany)
+  const contactInstance = _transformContact(timestamp, contact)
+
+  // NOTE: The order we let the Charging Module know about the change and persist the data to our own DB is important.
+  // In the legacy code we updated our DB first then let the Charging Module know. But if the CHA request failed it
+  // would appear to the user all was well; the address is updated in our service.
+  //
+  // Now we attempt to update the CHA first. We only attempt to update our DB if it succeeded (it will throw an error if
+  // it doesn't). If the CHA fails the user will know. If the CHA succeeds but our update fails the user will still see
+  // an error and try again. It matters not that we send the same information to the CHA; either way it will overwrite
+  // what is either in it or SOP. But this way the user is never left believing all is well when something has failed.
+  await SendCustomerChangeService.go(invoiceAccount, addressInstance, companyInstance, contactInstance)
+
+  return _persist(timestamp, invoiceAccount, addressInstance, companyInstance, contactInstance)
+}
+
+async function _fetchInvoiceAccount (invoiceAccountId) {
+  return InvoiceAccountModel.query()
+    .findById(invoiceAccountId)
+    .withGraphFetched('company')
+    .modifyGraph('company', (builder) => {
+      builder.select([
+        'name'
+      ])
+    })
+}
+
+/**
+ * Persist the changes to the WRLS database
+ *
+ * Any changes that need to be made to the WRLS DB are done here. We do them in a
+ * {@link https://vincit.github.io/objection.js/guide/transactions.html | transaction} object to ensure they either
+ * all get applied, or none do (no partial updates here!)
+ *
+ * We persist the address, company and contact first because we need their IDs in order to create the new
+ * `crm_v2.invoice_account_address` record. When we create that record we also need to apply an end date to any existing
+ * invoice account addresses with a null end date. This is how the service determines which address details are current
+ * (end date is null).
+ *
+ * The object we return has all 3 entities whether they were persisted or not. This gets passed back to the UI via
+ * the controller and is what it needs to then be able to redirect the user to the correct page.
+ *
+ * @param {Date} timestamp the timestamp to be used for any date created or updated values when persisting
+ * @param {module:InvoiceAccountModel} invoiceAccount the invoice (billing) account having its address changed
+ * @param {module:AddressModel} address the new address to be persisted (expected to be populated)
+ * @param {module:CompanyModel} company the new agent company to be persisted (not expected to be populated)
+ * @param {module:ContactModel} contact the new contact to be persisted (not expected to be populated)
```

Comment:
It took me a little figuring out to realise that when we say "expected to be populated" and "not expected to be populated", it's because `agentCompany` and `contact` are optional when calling the service but we always transform the day we receive (empty or not) into `companyInstance` and `contactInstance`. A little note in the comments to note why those two may not be populated might be useful.

Reply:
Added extra comments. Hopefully that helps.